src/container/srv_container.c

```diff
@@ -182,19 +182,56 @@ ds_cont_svc_fini(struct cont_svc **svcp)
 static int cont_svc_ec_agg_leader_start(struct cont_svc *svc);
 static void cont_svc_ec_agg_leader_stop(struct cont_svc *svc);
 
-void
+int
 ds_cont_svc_step_up(struct cont_svc *svc)
 {
-	int rc;
+	struct rdb_tx	tx;
+	d_iov_t		value;
+	uint32_t	version;
+	int		rc;
+
+	rc = rdb_tx_begin(svc->cs_rsvc->s_db, svc->cs_rsvc->s_term, &tx);
+	if (rc != 0)
+		goto out;
+	ABT_rwlock_rdlock(svc->cs_lock);
+
+	/* Check the layout version. */
+	d_iov_set(&value, &version, sizeof(version));
+	rc = rdb_tx_lookup(&tx, &svc->cs_root, &ds_cont_prop_version, &value);
+	if (rc == -DER_NONEXIST) {
+		D_ERROR(DF_UUID": incompatible layout version\n",
+			DP_UUID(svc->cs_pool_uuid));
+		rc = -DER_DF_INCOMPT;
+		goto out_lock;
+	} else if (rc != 0) {
+		D_ERROR(DF_UUID": failed to look up layout version: "DF_RC"\n",
+			DP_UUID(svc->cs_pool_uuid), DP_RC(rc));
+		goto out_lock;
+	}
+	if (version < DS_CONT_MD_VERSION_LOW || version > DS_CONT_MD_VERSION) {
+		D_ERROR(DF_UUID": incompatible layout version: %u not in "
+			"[%u, %u]\n", DP_UUID(svc->cs_pool_uuid), version,
+			DS_CONT_MD_VERSION_LOW, DS_CONT_MD_VERSION);
```

Comment:
I suggest logging a ras event here.  I added vos_report_layout_incompat (include/daos_srv/vos.h) which logs as

incompatible [type] may not be opened. Version %d is outside acceptable range %d-%d

Reply:
@jolivier23, RAS notifications were supposed to be left to future patches. Shouldn't vos use vos_durable_format_incompatible instead of pool_durable_format_incompatible? And shall I use pool_... for Pool Service, cont_... for Container Service, and rdb_... for RDB?

Reply:
@liw It's fine to leave to future patches.   I created a VOS API but it was intended to handle other cases as the first argument specifies what is incompatible.  I'd be fine changing the name of the API and relocating it if that is better but the intention was to have something that could handle the ras event for all of these layout incompatibility notifications.   Since VOS typically sits below all of these things, it seemed reasonable to put it there.

Reply:
@johannlombardi, @tanabarr, what should we do? Keep

    RAS_POOL_DF_INCOMPAT
    RAS_CONT_DF_INCOMPAT
    RAS_RDB_DF_INCOMPAT

or use

    RAS_POOL_DF_INCOMPAT

for all layout incompat RAS events?

I prefer the former because those events are different programmatically (via a number instead of substring in the message). Also, there may be layout incompat events from places outside pools (e.g., per-server metadata).

Reply:
@jolivier23, I think the only potentially common logic among pool, cont, rdb, and vos layout incompat events is the message format, to some degree. For rdb, I think it's good to have the file path and DB UUID in the message; for vos, we should add the path or the number in "vos-3", for instance. Given all these, it's okay for use ds_notify_ras_eventf directly, no?

Reply:
I think having separate event IDs makes sense and helps to differentiate without having to parse a message. When discussing the requirements for RAS, @johannlombardi expressed a preference for unique IDs of low granularity (preferably only one code path for raising a given event ID), with this in mind the former suggestion fits the bill and we shouldn't be hesitant about adding iDs.

Reply:
OK, I'll proceed to request landing then.